File: lib/data/subscription/hydra_subscription_uri.dart
/// URL handling for encrypted Hydra Subscription v2 sources.
///
/// Cryptography and JWE validation belong exclusively to HydraCore. The client
/// only extracts the out-of-band base64url key from `#hydra-key=...` and strips
/// the fragment before issuing an HTTP request.
class HydraSubscriptionUri {
  HydraSubscriptionUri._();

  static const plaintextMediaType = 'application/vnd.hydra.subscription+json';
  static const encryptedMediaType = 'application/jose+json';
  static const keyFragmentName = 'hydra-key';

  static String? keyFromUri(Uri uri) {
    final values = _rawKeyFragmentValues(uri);
    if (values.length != 1) return null;
    try {
      final value = Uri.decodeQueryComponent(values.single);
      if (value.isEmpty ||
          value != value.trim() ||
          !RegExp(r'^[A-Za-z0-9_-]{43}$').hasMatch(value)) {
        return null;
      }
      return value;
    } on FormatException {
      return null;
    }
  }

  static bool hasKeyFragment(Uri uri) => _rawKeyFragmentValues(uri).isNotEmpty;

  static bool hasKeyQueryParameter(Uri uri) {
    if (!uri.hasQuery) return false;
    for (final member in uri.query.split(RegExp(r'[&;]'))) {
      final separator = member.indexOf('=');
      final rawName = separator < 0 ? member : member.substring(0, separator);
      String name;
      try {
        name = Uri.decodeQueryComponent(rawName);
      } on FormatException {
        name = rawName;
      }
      if (name.toLowerCase() == keyFragmentName) return true;
    }
    return false;
  }

  static Uri withoutSecretFragment(Uri uri) {
    if (!uri.hasFragment) return uri;
    final serialized = uri.toString();
    return Uri.parse(serialized.substring(0, serialized.indexOf('#')));
  }

  static List<String> _rawKeyFragmentValues(Uri uri) {
    final fragment = uri.fragment;
    if (fragment.isEmpty) return const [];
    final values = <String>[];
    for (final member in fragment.split('&')) {
      final separator = member.indexOf('=');
      final rawName = separator < 0 ? member : member.substring(0, separator);
      String name;
      try {
        name = Uri.decodeQueryComponent(rawName);
      } on FormatException {
        name = rawName;
      }
      if (name != keyFragmentName) continue;
      values.add(separator < 0 ? '' : member.substring(separator + 1));
    }
    return values;
  }
}

File: core/subscription/src/commonMain/kotlin/io/hydrabox/core/subscription/HydraSubscriptionUri.kt
package io.hydrabox.core.subscription

/**
 * URL handling for encrypted Hydra Subscription v2 sources.
 *
 * Cryptography and JWE validation belong exclusively to the core. The client's only job
 * here is to lift the out-of-band key out of the `#hydra-key=` fragment and strip that
 * fragment before any request goes out — a fragment never leaves the device, a query
 * parameter does, which is why a key found in the query is rejected outright rather than
 * quietly accepted.
 */
object HydraSubscriptionUri {
    const val PLAINTEXT_MEDIA_TYPE = "application/vnd.hydra.subscription+json"
    const val ENCRYPTED_MEDIA_TYPE = "application/jose+json"
    const val KEY_FRAGMENT_NAME = "hydra-key"

    private val keyPattern = Regex("^[A-Za-z0-9_-]{43}$")

    /** The 32-byte key as base64url, or null when the source carries none. */
    fun keyOf(url: String): String? {
        val values = fragmentValues(url)
        if (values.size != 1) return null
        return values.single().takeIf { keyPattern.matches(it) }
    }

    fun hasKeyFragment(url: String): Boolean = fragmentValues(url).isNotEmpty()

    /** True when the key was put where it would be sent to the server. */
    fun hasKeyQueryParameter(url: String): Boolean {
        val query = url.substringBefore('#').substringAfter('?', "")
        if (query.isEmpty()) return false
        return query.split('&', ';').any { member ->
            member.substringBefore('=').lowercase() == KEY_FRAGMENT_NAME
        }
    }

    /** The URL as it may be requested: never carrying the key. */
    fun withoutSecretFragment(url: String): String = url.substringBefore('#')

    private fun fragmentValues(url: String): List<String> {
        val fragment = url.substringAfter('#', "")
        if (fragment.isEmpty()) return emptyList()
        return fragment.split('&').mapNotNull { member ->
            val separator = member.indexOf('=')
            val name = if (separator < 0) member else member.substring(0, separator)
            if (name != KEY_FRAGMENT_NAME) null else if (separator < 0) "" else member.substring(separator + 1)
        }
    }
}
